tripy/tests/spec_verification/test_dtype_constraints.py

```diff
@@ -15,120 +15,240 @@
+All other types do not have defaults and must be passed to the verifier using default_constraints_all.
+"""
+default_constraints_all = {
+    "__rtruediv__": {"self": 1},
+    "__rsub__": {"self": 1},
+    "__radd__": {"self": 1},
+    "__rpow__": {"self": 1},
+    "__rmul__": {"self": 1},
+    "softmax": {"dim": 1},
+    "concatenate": {"dim": 0},
+    "expand": {"sizes": tp.Tensor([3, 4]), "input": tp.ones((3, 1))},
+    "full": {"shape": tp.Tensor([3]), "value": 1},
+    "full_like": {"value": 1},
+    "flip": {"dim": 1},
+    "gather": {"dim": 0, "index": tp.Tensor([1])},
+    "iota": {"shape": tp.Tensor([3])},
+    "__matmul__": {"self": tp.ones((2, 3))},
+    "transpose": {"dim0": 0, "dim1": 1},
+    "permute": {"perm": [1, 0]},
+    "quantize": {"scale": tp.Tensor([1, 1, 1]), "dim": 0},
+    "sum": {"dim": 0},
+    "all": {"dim": 0},
+    "any": {"dim": 0},
+    "max": {"dim": 0},
+    "prod": {"dim": 0},
+    "mean": {"dim": 0},
+    "var": {"dim": 0},
+    "argmax": {"dim": 0},
+    "argmin": {"dim": 0},
+    "reshape": {"shape": tp.Tensor([6])},
+    "squeeze": {"input": tp.ones((3, 1)), "dims": (1)},
+    "__getitem__": {"index": 2},
+    "split": {"indices_or_sections": 2},
+    "unsqueeze": {"dim": 1},
+    "masked_fill": {"value": 1},
+    "ones": {"shape": tp.Tensor([3, 2])},
+    "zeros": {"shape": tp.Tensor([3, 2])},
+    "arange": {"start": 0, "stop": 5},
+}
+
+
+class InputValues:
+    dtype: str = None
+    init = None
+
+
+# Add default_constraints to input_values within TYPE_VERIFICATION
+for func_name, (
+    func_obj,
+    inputs_dict,
+    dtype_exceptions,
+    return_dtype,
+    dtype_variables,
+    dtype_constraints,
+) in TYPE_VERIFICATION.items():
+    # Complete the rest of the processing for TEST_VERIFICATION:
+    # Get names and type hints for each param.
+    func_sig = inspect.signature(func_obj)
+    param_dict = func_sig.parameters
+    # Construct inputs_dict.
+    for param_name, param_type in param_dict.items():
+        input_values = InputValues()
+        # If parameter had a default then use it otherwise skip.
+        if param_type.default is not param_type.empty:
+            # Checking if not equal to None since we want to enter if default is 0 or similar.
+            if param_type.default != None:
+                input_values.init = param_type.default
+        param_type = param_type.annotation
+        # If type is an optional or union get the first type.
+        while get_origin(param_type) in [Union, Optional]:
+            param_type = get_args(param_type)[0]
+            # ForwardRef refers to any case where type hint is a string.
+            if isinstance(param_type, ForwardRef):
+                param_type = param_type.__forward_arg__
+        # Check if there are any provided constraints and add them to the constraint dict.
+        input_values.dtype = dtype_constraints.get(param_name, None)
+        inputs_dict[param_name] = {param_type: input_values}
```

Comment:
Would it be better to always initialize the values here? Right now it's a bit mixed where if the value is provided in `default_constraints_all`, then it's intialized, but otherwise we just store the `dtype` constraint and do initialization later. 

If you always initialized here, you could remove the `InputValues` class and the `object_builders.py` file.

Reply:
Everything in `default_constraints_all` must still be casted to the right type and I currently do all final casting and initialization with each individual pytest test case so that a single error in initialization doesn't result in the whole test suite failing. I can also move everything from `object_builders.py` to `test_dtype_constraints.py`, but I believe that would lower readability.